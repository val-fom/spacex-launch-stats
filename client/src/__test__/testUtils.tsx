import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { MemoryRouter as Router } from 'react-router-dom';

// TODO: cosider to remove it
export function renderWithRouter(
  ui: React.ReactElement,
  { route = '/' } = {}
): RenderResult {
  const Wrapper: React.FC<{}> = ({ children }) => (
    <Router initialEntries={[route]}>{children}</Router>
  );
  return render(ui, { wrapper: Wrapper });
}
