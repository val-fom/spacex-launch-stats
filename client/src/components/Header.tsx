import React from 'react';
import { Header as SHeader } from 'semantic-ui-react';
import { Link } from 'react-router-dom';

export const Header: React.FC<{ style?: React.CSSProperties }> = ({
  style,
}) => (
  <SHeader data-testid="header" as="h1" textAlign="center" style={style}>
    <Link to="/">SpaceX</Link>
  </SHeader>
);
