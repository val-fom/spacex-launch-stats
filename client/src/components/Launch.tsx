/* eslint-disable camelcase */
/* eslint-disable @typescript-eslint/camelcase */
import React from 'react';
import { useQuery } from '@apollo/react-hooks';
import { gql } from 'apollo-boost';
import { Table, Icon } from 'semantic-ui-react';

export const GET_LAUNCH = gql`
  query Launch($flight_number: Int!) {
    launch(flight_number: $flight_number) {
      flight_number
      mission_name
      launch_year
      launch_date_local
      launch_success
      rocket {
        rocket_id
        rocket_name
        rocket_type
      }
    }
  }
`;

export const Launch: React.FC<{ flightNumber?: string }> = ({
  flightNumber,
}) => {
  const { loading, error, data } = useQuery<{ launch: Launch }>(GET_LAUNCH, {
    variables: {
      flight_number: flightNumber ? parseInt(flightNumber) : undefined,
    },
  });

  if (loading) return <p data-testid="launch-loading">Loading...</p>;

  if (error) return <p data-testid="launch-error">Error</p>;

  if (!data?.launch) return <p>No data</p>;

  const {
    flight_number,
    mission_name,
    launch_year,
    launch_date_local,
    launch_success,
    rocket: { rocket_id, rocket_name, rocket_type },
  } = data.launch;

  return (
    <Table celled unstackable>
      <Table.Body>
        <Table.Row>
          <Table.Cell>Flight number</Table.Cell>
          <Table.Cell>{flight_number}</Table.Cell>
        </Table.Row>
        <Table.Row>
          <Table.Cell>Mission name</Table.Cell>
          <Table.Cell>{mission_name}</Table.Cell>
        </Table.Row>
        <Table.Row>
          <Table.Cell>Launch year</Table.Cell>
          <Table.Cell>{launch_year}</Table.Cell>
        </Table.Row>
        <Table.Row>
          <Table.Cell>Launch date local</Table.Cell>
          <Table.Cell>{launch_date_local}</Table.Cell>
        </Table.Row>
        <Table.Row>
          <Table.Cell>Rocket id</Table.Cell>
          <Table.Cell>{rocket_id}</Table.Cell>
        </Table.Row>
        <Table.Row>
          <Table.Cell>Rocket name</Table.Cell>
          <Table.Cell>{rocket_name}</Table.Cell>
        </Table.Row>
        <Table.Row>
          <Table.Cell>Rocket type</Table.Cell>
          <Table.Cell>{rocket_type}</Table.Cell>
        </Table.Row>
        <Table.Row>
          <Table.Cell>Launch success</Table.Cell>
          <Table.Cell>
            {launch_success ? (
              <Icon name="checkmark" color="green" />
            ) : (
              <Icon name="close" color="red" />
            )}
          </Table.Cell>
        </Table.Row>
      </Table.Body>
    </Table>
  );
};
