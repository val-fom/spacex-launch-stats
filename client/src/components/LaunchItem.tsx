import React from 'react';
import { Link } from 'react-router-dom';

export const LaunchItem: React.FC<{ launch: Launch }> = ({ launch }) => (
  <div className="App-launch">
    <Link to={`/launch/${launch.flight_number}`}>
      <div>flight_number</div>
      <div>{launch.flight_number}</div>
    </Link>
    <div>launch_date_local</div>
    <div>{launch.launch_date_local}</div>
    <div>mission_name</div>
    <div>{launch.mission_name}</div>
    <div>rocket_name</div>
    <div>{launch.rocket.rocket_name}</div>
  </div>
);
