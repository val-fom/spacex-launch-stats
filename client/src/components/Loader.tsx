import React from 'react';
import { Loader as SLoader } from 'semantic-ui-react';

export const Loader: React.FC<{}> = () => (
  <SLoader active inline="centered" size="big" data-testid="loader" />
);
