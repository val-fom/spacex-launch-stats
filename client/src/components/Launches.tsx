/* eslint-disable camelcase */
import React from 'react';
import { useQuery } from '@apollo/react-hooks';
import { gql } from 'apollo-boost';
import { Card } from 'semantic-ui-react';
import { Loader } from 'components/Loader';
import { LaunchItem } from './LaunchItem';

export const GET_LAUNCHES = gql`
  query Launches {
    launches {
      flight_number
      mission_name
      launch_year
      launch_date_local
      launch_success
      rocket {
        rocket_id
        rocket_name
        rocket_type
      }
    }
  }
`;

export const Launches: React.FC<{}> = () => {
  const { loading, error, data } = useQuery<{ launches: Launch[] }>(
    GET_LAUNCHES
  );

  if (loading) return <Loader />;

  if (error) return <p>Error</p>;

  return (
    <Card.Group
      data-testid="launches-data"
      itemsPerRow={window.innerWidth > 767 ? 2 : 1}
    >
      {data?.launches.map(launch => (
        <LaunchItem
          key={`${launch.flight_number}${launch.mission_name}`}
          launch={launch}
        />
      ))}
    </Card.Group>
  );
};
