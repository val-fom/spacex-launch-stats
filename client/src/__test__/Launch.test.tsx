/* eslint-disable @typescript-eslint/camelcase */
import React from 'react';
import { MockedProvider } from '@apollo/react-testing';
import { Launch, LAUNCH } from '../components/Launch';
import { renderWithRouter } from './testUtils';

const mocks = [
  {
    request: {
      query: LAUNCH,
      variables: {
        flight_number: 1,
      },
    },
    result: {
      data: {
        launch: {
          flight_number: 1,
          mission_name: 'FalconSat',
          launch_year: '2006',
          launch_date_local: '2006-03-25T10:30:00+12:00',
          launch_success: false,
          rocket: {
            rocket_id: 'falcon1',
            rocket_name: 'Falcon 1',
            rocket_type: 'Merlin A',
          },
        },
      },
    },
  },
];

test('renders without crashes', () => {
  const { asFragment } = renderWithRouter(
    <MockedProvider mocks={mocks} addTypename={false}>
      <Launch />
    </MockedProvider>,
    { route: '/1' }
  );
  expect(asFragment()).toMatchSnapshot();
});
