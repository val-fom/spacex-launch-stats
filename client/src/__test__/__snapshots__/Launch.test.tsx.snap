// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashes 1`] = `
<DocumentFragment>
  <p>
    Loading...
  </p>
</DocumentFragment>
`;
